### ISSUE TITLE
FP16 Training of VAE

### ISSUE THREAD

**Issue by wangwhTj** [VAST-AI-Research/TripoSF#5]
I tried setting up a training pipeline where the local point encoder uses FP32 and the VAE encoder is preset to FP16. However, during training, the parameters in the FP16 layers (except for the output layers) are not updating. How can I resolve this?
Details:
`     optimizer = AdamW(model.parameters(), lr=args.lr, weight_decay=0.0, eps=1e-4)`
`      .......`
`            with torch.cuda.amp.autocast(dtype=torch.float16):`
`                posterior, recon_pcs_per_voxel = model(input_pcs, input_voxel)`
forward of model:
`    def forward(self, points_sample, sparse_voxel_coords):`
`        with torch.autocast("cuda", dtype=torch.float32):`
`            sparse_pc_features = self.local_pc_encoder(points_sample, sparse_voxel_coords, res=self.cfg.resolution, bbox_size=(-0.5, 0.5))`
`        sparse_tensor = sp.SparseTensor(sparse_pc_features, sparse_voxel_coords)`
`        # with torch.autocast("cuda", dtype=torch.float16):`
`        latent, posterior = self.encoder(sparse_tensor, sample_posterior=False)`
`        recon_pcs_per_voxel = self.decoder(latent)`
`        return posterior, recon_pcs_per_voxel`

**Comment by TianhaoZhao668** [comment]
Hello, could you mind sharing your training code if possible so that we can fix it together？

**Comment by jingyu198** [comment]
Hi, I have the same problem, and cause NAN loss during training. do you solve this?

**Comment by wangwhTj** [comment]
@jingyu198 I compare tripo's project with TRELLIS, and find the architecture of their models files are similar. So I use the training script from TRELLIS and now it works. (I guess this project is built upon TRELLIS's project)

**Comment by jingyu198** [comment]
Thanks! I will try TRELLIS too

**Comment by TianhaoZhao668** [comment]
> Hi, I have the same problem, and cause NAN loss during training. do you solve this?

Use transformers' optimizer maybe OK  instead of torch's

**Comment by jingyu198** [comment]
> Use transformers' optimizer maybe OK instead of torch's

Thanks, bro. This works for me!

**Comment by yipingp** [comment]
Hi! I'm working on training as well. I'm wondering if you encountered the mesh extraction from flexicube kept failing at the beginning of training? Thanks!

**Comment by jingyu198** [comment]
> Hi! I'm working on training as well. I'm wondering if you encountered the mesh extraction from flexicube kept failing at the beginning of training? Thanks!

this does not happen in the begining,  there may be something wrong with your trianing code

**Comment by jingyu198** [comment]
> Hi! I'm working on training as well. I'm wondering if you encountered the mesh extraction from flexicube kept failing at the beginning of training? Thanks!

even with the most basic losses without pruning loss and local rendering, the training process seems smooth. at least there is not so much change in the visualization result.

**Comment by yipingp** [comment]
> even with the most basic losses without pruning loss and local rendering, the training process seems smooth. at least there is not so much change in the visualization result.

Hi, thanks! Do you train from scratch or finetune? Can you get valid mesh at the beginning of training, without any sdf initialization?

**Comment by jingyu198** [comment]
> a

I am finetuning the model, so at the beggining there are valid meshes

**Comment by jingyu198** [comment]
> > even with the most basic losses without pruning loss and local rendering, the training process seems smooth. at least there is not so much change in the visualization result.
> 
> Hi, thanks! Do you train from scratch or finetune? Can you get valid mesh at the beginning of training, without any sdf initialization?

I am finetuning the model, so at the beggining there are valid meshes

**Comment by XinLeiChenff** [comment]
Can you train at 128 resolution, or even 256 and above? It seems their paper proposes a partial-voxel training strategy, whereas Trellis is quite different—Trellis generates the full set of voxels for rendering-based training. If we use Trellis’s training scheme, it would likely cause an OOM at high resolutions, right?

> [@jingyu198](https://github.com/jingyu198) I compare tripo's project with TRELLIS, and find the architecture of their models files are similar. So I use the training script from TRELLIS and now it works. (I guess this project is built upon TRELLIS's project)

**Comment by fangchuan** [comment]
Same question @XinLeiChenff . This code base doesnot offer partial-voxel training, which is critical for  training in high-res,

**Comment by animemory** [comment]
> [@jingyu198](https://github.com/jingyu198) I compare tripo's project with TRELLIS, and find the architecture of their models files are similar. So I use the training script from TRELLIS and now it works. (I guess this project is built upon TRELLIS's project)

hi, do u mean the vae encoder training or decoder training?
